integration_test.go

```diff
@@ -0,0 +1,160 @@
+package main
+
+import (
+	"bufio"
+	"bytes"
+	"net"
+	"net/smtp"
+	"net/textproto"
+	"os"
+	"strings"
+	"testing"
+	"time"
+
+	"github.com/chrj/smtpd"
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+)
+
+type testSMTPServer struct {
+	msgs *[]smtpd.Envelope
+	addr string
+}
+
+func startTestSMTPServer(t *testing.T) *testSMTPServer {
+	t.Helper()
+
+	msgs := &[]smtpd.Envelope{}
+	srv := &smtpd.Server{
+		ConnectionChecker: func(peer smtpd.Peer) error {
+			t.Logf("Connection from %s", peer.HeloName)
+			return nil
+		},
+		HeloChecker: func(peer smtpd.Peer, name string) error {
+			t.Logf("HELO (%s) %s", peer.Protocol, name)
+			return nil
+		},
+		SenderChecker: func(peer smtpd.Peer, addr string) error {
+			t.Logf("MAIL FROM %s", addr)
+			return nil
+		},
+		RecipientChecker: func(peer smtpd.Peer, addr string) error {
+			t.Logf("RCPT TO %s", addr)
+			return nil
+		},
+		Handler: func(peer smtpd.Peer, env smtpd.Envelope) error {
+			t.Logf("DATA\n----\n%s\n----", env.Data)
+			m := append(*msgs, env)
+			*msgs = m
+			return nil
+		},
+	}
+
+	l, err := net.Listen("tcp", "127.0.0.1:0")
```

Comment:
I vaguely remember a convo on slack about this. This listens for the connection on **any** available port right?

Reply:
nvm, comment on line 94 confirms this